src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,147 @@
+        } else {
+            tail.right = newNode;
+            newNode.left = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (size != index) {
+            checkIndex(index);
+        }
+        Node<T> newNode = new Node<>(value);
+        size++;
+        if (this.isEmpty() && index == 0) {
+            head = tail = newNode;
+            return;
+        }
+        if (index == 0) {
+            newNode.right = head;
+            if (head != null) {
+                head.left = newNode;
+            }
+            head = newNode;
+            return;
+        }
+        if (index == size - 1) {
+            tail.right = newNode;
+            newNode.left = tail;
+            tail = newNode;
+            return;
+        }
+        Node<T> tempNode = findByIndex(index);
+        tempNode.left.right = newNode;
+        newNode.right = tempNode;
+        newNode.left = tempNode.left;
+        tempNode.left = newNode;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            this.add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> temp = findByIndex(index);
+        return temp.data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> temp = findByIndex(index);
+        T old = temp.data;
+        temp.data = value;
+        return old;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> temp = findByIndex(index);
+        if (temp == head) {
+            if (this.size > 1) {
+                head = temp.right;
+                temp.right = null;
+                head.left = null;
+            } else {
+                head = null;
+                tail = null;
+            }
+        } else if (temp == tail) {
+            temp.left.right = null;
+        } else {
+            temp.left.right = temp.right;
+            temp.right.left = temp.left;
+        }
+        size--;
+        return temp.data;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> temp = head;
+        for (int i = 0; i < size; i++) {
+            if ((temp.data != null && temp.data.equals(object))
+                    || (temp.data == null && null == object)) {
+                this.remove(i);
+                return true;
+            }
+            temp = temp.right;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return head == null;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index > size - 1 || (this.isEmpty() && index > 0)) {
```

Comment:
create other method to check index for add method
`if (index < 0 || index > size) {`

```suggestion
        if (index < 0 || index >= size) {
```